yarn-project/end-to-end/src/e2e_key_rotation.test.ts

```diff
@@ -0,0 +1,197 @@
+    const isUserSynchronized = async () => {
+      return await wallet.isAccountStateSynchronized(owner);
+    };
+    await retryUntil(isUserSynchronized, `synch of user ${owner.toString()}`, 10);
+  };
+
+  const crossDelay = async () => {
+    for (let i = 0; i < SHARED_MUTABLE_DELAY; i++) {
+      // We send arbitrary tx to mine a block
+      await testContract.methods.emit_unencrypted(0).send().wait();
+    }
+  };
+
+  const expectTokenBalance = async (
+    wallet: Wallet,
+    tokenAddress: AztecAddress,
+    owner: AztecAddress,
+    expectedBalance: bigint,
+    checkIfSynchronized = true,
+  ) => {
+    if (checkIfSynchronized) {
+      // First wait until the corresponding PXE has synchronized the account
+      await awaitUserSynchronized(wallet, owner);
+    }
+
+    // Then check the balance
+    const contractWithWallet = await TokenContract.at(tokenAddress, wallet);
+    const balance = await contractWithWallet.methods.balance_of_private(owner).simulate({ from: owner });
+    logger.info(`Account ${owner} balance: ${balance}`);
+    expect(balance).toBe(expectedBalance);
+  };
+
+  const deployTokenContract = async (initialAdminBalance: bigint, admin: AztecAddress, pxe: PXE) => {
+    logger.info(`Deploying Token contract...`);
+    const contract = await TokenContract.deploy(walletA, admin, 'TokenName', 'TokenSymbol', 18).send().deployed();
+
+    if (initialAdminBalance > 0n) {
+      await mintTokens(contract, admin, initialAdminBalance, pxe);
+    }
+
+    logger.info('L2 contract deployed');
+
+    return contract.instance;
+  };
+
+  const mintTokens = async (contract: TokenContract, recipient: AztecAddress, balance: bigint, pxe: PXE) => {
+    const secret = Fr.random();
+    const secretHash = computeSecretHash(secret);
+
+    const receipt = await contract.methods.mint_private(balance, secretHash).send().wait();
+
+    const note = new Note([new Fr(balance), secretHash]);
+    const extendedNote = new ExtendedNote(
+      note,
+      recipient,
+      contract.address,
+      TokenContract.storage.pending_shields.slot,
+      TokenContract.notes.TransparentNote.id,
+      receipt.txHash,
+    );
+    await pxe.addNote(extendedNote);
+
+    await contract.methods.redeem_shield(recipient, balance, secret).send().wait();
+  };
+
+  it(`We test key rotation in four steps.
+    1. We transfer funds from A to B.
+    2. We rotate B's keys by calling rotateMasterNullifierKey on our wallet and by calling rotate_npk_m on the Key Registry.
+    3. After the key rotation has been applied, we then transfer more funds from A to B.
+    4. We finally send all of the funds back from B to A, proving that all the funds sent before and after key rotation are spendable,
+    and that we are able to spend notes associated with different nullifying keys`, async () => {
+    const initialBalance = 987n;
+    const transferAmount1 = 654n;
+    // const transferAmount2 = 323n;
```

Comment:
Can remove this.

Reply:
Nuked this in https://github.com/AztecProtocol/aztec-packages/pull/6405/commits/968085851a7b24561726329b7310238d9e4c2fc9